librad/src/net/protocol/event.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright © 2019-2020 The Radicle Foundation <hello@radicle.foundation>
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+use std::net::SocketAddr;
+
+use super::{broadcast, gossip, membership};
+use crate::PeerId;
+
+#[derive(Clone)]
+pub enum Downstream {
+    Gossip(downstream::Gossip),
+    Info(downstream::Info),
+}
+
+pub mod downstream {
+    use super::*;
+
+    use std::sync::Arc;
+
+    use parking_lot::Mutex;
+    use tokio::sync::oneshot;
+
+    pub type Reply<T> = Arc<Mutex<Option<oneshot::Sender<T>>>>;
+
+    #[derive(Clone, Debug)]
+    pub enum Gossip {
+        Announce(gossip::Payload),
+        Query(gossip::Payload),
+    }
+
+    impl Gossip {
+        pub fn payload(self) -> gossip::Payload {
+            match self {
+                Self::Announce(p) => p,
+                Self::Query(p) => p,
+            }
+        }
+    }
+
+    #[derive(Clone)]
+    pub enum Info {
+        ConnectedPeers(Reply<Vec<PeerId>>),
+        Stats(Reply<Stats>),
+    }
+
+    #[derive(Clone, Debug, Default)]
+    pub struct Stats {
+        pub connections_total: usize,
+        pub connected_peers: usize,
+        pub membership_active: usize,
+        pub membership_passive: usize,
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum Upstream {
+    Endpoint(upstream::Endpoint),
+    Gossip(Box<upstream::Gossip<SocketAddr, gossip::Payload>>),
+    Membership(membership::Transition<SocketAddr>),
+}
+
+pub mod upstream {
+    use super::*;
+
+    use std::time::Duration;
+
+    use futures::{future, FutureExt as _, StreamExt as _, TryStreamExt as _};
+    use futures_timer::Delay;
+    use thiserror::Error;
+
+    use crate::net::protocol::{PeerInfo, RecvError};
+
+    #[derive(Clone, Debug)]
+    pub enum Endpoint {
+        Up { listen_addrs: Vec<SocketAddr> },
+        Down,
+    }
+
+    impl From<Endpoint> for Upstream {
+        fn from(e: Endpoint) -> Self {
+            Self::Endpoint(e)
+        }
+    }
+
+    #[derive(Clone, Debug)]
+    pub enum Gossip<Addr, Payload>
+    where
+        Addr: Clone + Ord,
+    {
+        Applied {
```

Comment:
This did not change, so I'm surprised by the comment. Would you mind painting your pink, as I am having trouble to see what confuses you?

Reply:
Looking at [`maint`](https://github.com/radicle-dev/radicle-link/blob/285bdfee6956084c955c079a3789b4ac6468f7d2/librad/src/net/peer.rs#L113-L117) it was a field in `FetchInfo`. The only mention of `Applied` there was in `PutResult` but here we have `PutResult` and this `enum Gossip`.

My confusion came while reviewing the PR, seeing `Applied` and thinking it was to do with `PutResult`. But then I realised that there were two different enums. I then found it slightly weird that the `Gossip` value could be `Gossip::Applied` + `PutResult::Unintersting` or `Gossip::Applied` + `PutResult::Error`.

On reflection though that's not a big deal because we have the `enum` namespacing, i.e. `Gossip::Applied` vs `PutResult::Applied` and so we can understand the context from that.

Reply:
I see, thanks. There is a bit of a disentaglement of broadcasting (aka "gossip") and membership going on. Also, much to my dismay, we have to embrace this enterprise tin can phone API now. Conceivably, we might want to publish other types of gossip-related events on the enterprise service bus in the future, so I needed a variant name which denotes when this thing is to be expected.

Perhaps `Put`, despite the ambiguous tense?

Reply:
Ya `Put` makes sense to me -- it can be directly linked with the result of calling `LocalStorage::put`.